langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java

```diff
@@ -109,14 +106,81 @@ static Object[] prepareArguments(
             if (argumentsMap.containsKey(parameterName)) {
                 Object argument = argumentsMap.get(parameterName);
                 Class<?> parameterType = parameters[i].getType();
-
-                arguments[i] = coerceArgument(argument, parameterName, parameterType);
+                if( parameterType == List.class || parameterType == Set.class || parameterType == Map.class)  {
+                    ParameterizedType parameterizedType = (ParameterizedType) parameters[i].getParameterizedType();
+                    arguments[i] = coerceCollectionArgument(argument, parameterName, parameterType, parameterizedType);
+                } if(parameterType.isArray() && argument instanceof Collection) {
+                    arguments[i] = coerceCollectionArgument(argument, parameterName, parameterType, parameterType.getComponentType());
+                }else {
+                    arguments[i] = coerceArgument(argument, parameterName, parameterType);
+                }
             }
         }
 
         return arguments;
     }
 
+    @SneakyThrows
+    static Object coerceCollectionArgument(
+            Object argument,
+            String parameterName,
+            Class<?> parameterType,
+            Object parameterizedType
+    )  {
+        String json = Json.toJson(argument);
+        if (parameterType == List.class) {
+            List objects = Json.fromJson(json , List.class);
+            Class<?> elementType  = Class.forName( ((ParameterizedType)parameterizedType).getActualTypeArguments()[0].getTypeName());
```

Comment:
Is there a way to avoid calling `Class.forName()`?

Reply:
Nope. But it should not fail. Because, It would have failed much before, if the class was not present.